src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/HashPartitioningItemMapper.java

```diff
@@ -172,188 +122,40 @@ private int getPrimitiveValueHashCode(ScalarAttributeType type, AttributeValue v
 
         Map<String, AttributeValue> virtualItem = new HashMap<>(physicalItem);
 
-        // move table primary key fields
-        removePhysicalKeys(virtualItem, virtualTable.getPrimaryKey(), physicalTable.getPrimaryKey(), true);
+        // remove physical table PK fields and add back virtual fields if needed
+        reversePhysicalKeys(virtualItem, virtualTable.getPrimaryKey(), physicalTable.getPrimaryKey(), true);
 
-        // copy index fields
+        // remove physical secondary index fields and add back virtual fields if needed
         for (DynamoSecondaryIndex virtualSi : virtualTable.getSis()) {
             DynamoSecondaryIndex physicalSi = lookUpPhysicalSecondaryIndex.apply(virtualSi);
-            removePhysicalKeys(virtualItem, virtualSi.getPrimaryKey(), physicalSi.getPrimaryKey(), false);
+            reversePhysicalKeys(virtualItem, virtualSi.getPrimaryKey(), physicalSi.getPrimaryKey(), false);
         }
 
         return virtualItem;
     }
 
-    private void removePhysicalKeys(Map<String, AttributeValue> item, PrimaryKey virtualPK,
-                                    PrimaryKey physicalPK, boolean required) {
+    private void reversePhysicalKeys(Map<String, AttributeValue> item, PrimaryKey virtualPK,
+                                     PrimaryKey physicalPK, boolean required) {
         item.remove(physicalPK.getHashKey());
         AttributeValue physicalRkValue = item.remove(physicalPK.getRangeKey().get());
         if (required) {
             checkNotNull(physicalRkValue, "Physical item missing range key value");
         }
 
+        // item might already have virtual PK fields populated, but not if we're reversing physical key fields only
+        // (e.g., mapping a physical query's LastEvaluatedKey), so reverse physical key fields in case it's needed.
         if (physicalRkValue != null) {
+            ByteBuffer physicalRkBytes = physicalRkValue.getB();
+            physicalRkBytes.rewind();
             if (virtualPK.getRangeKey().isPresent()) {
-                AttributeValue[] vs = AttributeBytesConverter.fromBytes(virtualPK.getHashKeyType(),
-                    virtualPK.getRangeKeyType().get(), physicalRkValue.getB());
-                item.put(virtualPK.getHashKey(), vs[0]);
-                item.put(virtualPK.getRangeKey().get(), vs[1]);
+                AttributeValue[] virtualValues = fromBytes(virtualPK.getHashKeyType(),
+                    virtualPK.getRangeKeyType().get(), physicalRkBytes);
+                item.put(virtualPK.getHashKey(), virtualValues[0]);
+                item.put(virtualPK.getRangeKey().get(), virtualValues[1]);
             } else {
-                item.put(virtualPK.getHashKey(), AttributeBytesConverter.fromBytes(virtualPK.getHashKeyType(),
-                    physicalRkValue.getB()));
-            }
-        }
-    }
-
```

Comment:
The logic for converting virtual HK and RK values into physical HK and RK bytes is now in HashPartitioningKeyMapper (and its inner classes)